src/main/java/ninjabrainbot/model/actions/alladvancements/TryAddAllAdvancementsStructureAction.java

```diff
@@ -28,29 +47,49 @@ public void execute() {
 		if (structureType == StructureType.Unknown)
 			return;
 
-		StructurePosition structurePosition =
-				structureType == StructureType.Outpost
-						? getOutpostPosition(playerPosition)
-						: new StructurePosition((int) Math.floor(playerPosition.xInOverworld()), (int) Math.floor(playerPosition.zInOverworld()), playerPositionObservable);
+		StructurePosition structurePosition = getStructurePosition(structureType);
 
 		IDataComponent<StructurePosition> dataComponent = getDataComponentFromStructureType(structureType);
-		if (dataComponent.get() != null)
+		// Cities can be queried multiple times, so overwrite the position.
+		if (structureType != StructureType.CityQuery && dataComponent.get() != null)
 			return;
 
 		dataComponent.set(structurePosition);
 	}
 
 	private StructureType getAllAdvancementStructureTypeFromPlayerPosition(IDetailedPlayerPosition t) {
-		if (t.isInNether())
-			return StructureType.Unknown;
+		if (!preferences.oneDotTwentyPlusAA.get()) {
```

Comment:
There is too much nesting and too much repetition in this method. I think this would achieve the same result but with less repetition:

1. If in nether, return Unknown (this applies to both pre and post 1.20)
2. If in the end, return ShulkerTransport if 1.20+, or Unknown if pre 1.20
3. Check for spawn
4. Check for monument
5. If 1.20+, check for DeepDark and CityQuery
6. Return Outpost